R/delete_item.R

```diff
@@ -0,0 +1,89 @@
+#' Delete a ScienceBase item
+#' 
+#' Deletes one or more generic items, including their child items and files as
+#' applicable & specified.
+#' 
+#' @param item_ids character. SB id[s]
+#' @param item_names character. The name[s] by which to refer to the item[s] in 
+#'   messages
+#' @param delete_files logical. Should files be sought and deleted from within 
+#'   this item? If FALSE, files will be ignored.
+#' @param delete_children logical. Should child items be sought and recursively 
+#'   deleted from within this item? If FALSE, children will be ignored.
+#' @param delete_item logical. Should the item itself be deleted after files 
+#'   and/or children are deleted?
+#' @param verbose logical. Should status messages be given?
+#' @keywords internal
+delete_item <- function(item_ids, item_names, delete_files=FALSE, delete_children=FALSE, delete_item=TRUE, verbose=TRUE) {
+  
+  if(is.null(current_session())) stop("session is NULL. call sbtools::authenticate_sb() before deleting")
+  if(length(item_ids) != length(item_names)) stop("expecting item_ids and item_names to have the same length")
+  
+  # delete in a loop over the assumed-to-be parallel vectors item_ids and item_names
+  deletion_msgs <- lapply(1:length(item_ids), function(q) {
+    
+    item_id <- item_ids[q]
+    item_name <- item_names[q]
+    
+    # do nothing & return if it's already gone
+    if(is.na(item_id)) {
+      if(isTRUE(verbose)) message("skipping deletion of missing item ", item_name)
+      return(NA)
+    }
+    
+    # declare intent
+    if(verbose) message("deleting item ", item_name, " (", item_id, ")")
+    deletion_msg <- list()
+      
+    # if requested, delete all files from the item
+    if(isTRUE(delete_files)) {
+      # do the deletion
+      item_status <- item_rm_files(item_id)
+      # sleep to give time for full deletion
+      for(wait in 1:50) {
+        Sys.sleep(0.2)
+        if(nrow(item_list_files(item_id)) == 0) break
+        if(wait==50) stop("failed to delete files & therefore item (", item_id, ")")
+      }
+      deletion_msg <- c(deletion_msg, list("deleted all files")) # this will be overwritten if delete_item=TRUE
+    }
+    
+    # if requested, delete all children from the item. this is recursive - be careful!
+    if(isTRUE(delete_children)) {
+      # identify child items to delete
+      children <- item_list_children(item_id, limit=1000)
+      # delete them
+      if(nrow(children) > 0) {
+        for(child in children$id) {
+          delete_item(item_id=child, item_name=child, delete_files=TRUE, delete_children=TRUE, delete_item=TRUE, verbose=verbose)
+        }
+      }
+      # sleep to give time for full deletion
+      for(wait in 1:50) {
+        Sys.sleep(0.2)
```

Comment:
just a thought, but it might be nice for general retry logic to have an exponentially growing wait time that stops when it hits a certain value, which is parameterized. Since you do this a lot too, it might be worth pulling out as a separate "hold up" function that checks for state.